### PR TITLE
Adding Permissions

### PR DESCRIPTION
**Pull request by notgull**

### CODE REVIEW

File: backend/sources.mk [hunk 1]
```diff
@@ -23,4 +23,6 @@ BACKEND_SOURCES := \
 	backend/gulpfile.js \
 	backend/package.json \
 	backend/tsconfig.json \
-	backend/src/**/*
+	backend/src/* \
+	backend/src/**/* \
+	backend/src/**/**/*
```
Comment (emmiegit) [inline backend/sources.mk]: This doesn't do anything. Replace all of this with
```make
$(shell find backend src -name '*.ts' 2> /dev/null)
```

File: backend/gulpfile.js [hunk 2]
```diff
@@ -24,7 +24,7 @@ const ts = require('gulp-typescript');
 const tsProject = ts.createProject('tsconfig.json');
 
 gulp.task('default', () => (
-  gulp.src('src/**/*.ts')
+  gulp.src('src/**/**/*.ts')
```
Comment (emmiegit) [inline backend/gulpfile.js:27]: Does this work? The multi glob should either work or not -- consult the gulp docs.

File: backend/dist/app [hunk 3]
```diff
@@ -1 +1 @@
-.
```
Comment (emmiegit) [inline backend/dist/app:1]: ?

File: backend/src/index.ts [hunk 4]
```diff
@@ -197,6 +182,11 @@ app.post("/sys/process-login", function(req: express.Request, res: express.Respo
 
   // firstly, validate both whether the user exists and whether the password is correct
   User.loadByUsername(username).then((user: User) => {
+    if (!user) {
+      res.redirect(`/sys/login?errorCode=512`);
+      return;
+    }
```
Comment (emmiegit) [inline backend/src/index.ts:188]: Side note, I dislike the whole login error code redirect pattern, since the bitmask and actual error are very opaque to anybody reading the code.

File: backend/src/index.ts [hunk 5]
```diff
@@ -253,8 +243,11 @@ app.get("/sys/register", function(req: express.Request, res: express.Response) {
 });
 
 function onEmailVerify(username: string, pwHash: string, email: string): void {
-  User.createNewUser(username, email, pwHash).then(() => { console.log("Created user " + username); })
-    .catch((err: Error) => { console.error(`User creation error: ${err}`); });
+  Role.loadDefaultRole().then((role: Role) => {
+    User.createNewUser(username, email, pwHash, role).then(() => { 
+      console.log(`Created user ${username}`); 
+    }).catch((err: Error) => { console.error(`User creation error: ${err}`); });
+  }).catch((err: Error) => { console.error(`Error loading default role: ${err}`); });
```
Comment (emmiegit) [inline backend/src/index.ts]: Please convert to async/await instead of `.then()` chaining

File: backend/src/services/metadata/autocreate-404.ts [hunk 6]
```diff
@@ -22,71 +22,101 @@ import * as diff from 'diff';
...
+// an async way of going about this
+async function autocreateAsync(): Promise<number> {
+  // create roles for system, admin, and for default users
+  let fullPermission = 0xFFFFFFFF;
+  let systemRole = await createRole(Role.systemRoleName, fullPermission);
+  let defaultRole = await createRole(Role.defaultRoleName, null);
+  let adminRole = await createRole(Role.adminRoleName, fullPermission); 
+
+  // create system user
+  let systemUser = await User.createNewUser("system", "noreply@scipnet.net", "**DONTLOGINTOTHISACCOUNT**", systemRole, true);
```
Comment (emmiegit) [inline backend/src/services/metadata/autocreate-404.ts:67]: I've commented on the ad-hoc user creation before.

File: backend/src/services/pagereq/index.ts [hunk 7]
```diff
@@ -171,15 +210,29 @@ function removeEditLock(username: string, args: ArgsMapping, next: PRSCallback)
...
   let returnVal = genReturnVal();
 
+  if (!(user.hasPermission("editPages"))) {
+    return permissionDenied();
+  }
+
   let pMeta = await Metadata.load_by_slug(args.pagename);
 
+  // if the metadata is null, we are creating a new page. check for permission
+  if (!pMeta && !(user.hasPermission("createPages"))) {
```
Comment (emmiegit) [inline backend/src/services/pagereq/index.ts]: Why not
```suggestion
  if (!pMeta && !user.hasPermission("createPages")) {
```

File: backend/src/services/pagereq/index.ts [hunk 8]
```diff
@@ -171,15 +210,29 @@ function removeEditLock(username: string, args: ArgsMapping, next: PRSCallback)
...
+    return permissionDenied();
+  }
+
+  // if the page is locked and we don't have permission, return
+  if (pMeta && pMeta.locked_at && !(user.hasPermission("modifyLockedPages"))) {
```
Comment (emmiegit) [inline backend/src/services/pagereq/index.ts]: Ditto

File: backend/src/services/pagereq/index.ts [hunk 9]
```diff
@@ -456,32 +521,35 @@ export function request(name: string, username: string, args: ArgsMapping, next:
...
+  User.loadByUsername(username).then((user: Nullable<User>) => {
+    if ((!user) && username) { next(genErrorVal(new Error("Bad user id"))); return; }
+    if (!(user)) console.log("User.loadByUsername gave no matches...");
 
     if (!args["pagename"] || args["pagename"].length === 0)
       args["pagename"] = "main";
 
-    args['user_id'] = user_id;
+    args['user'] = user;
+    if (user)
```
Comment (emmiegit) [inline backend/src/services/pagereq/index.ts]: Please wrap all `if`s in braces

File: backend/src/services/pagereq/index.ts [hunk 10]
```diff
@@ -456,32 +521,35 @@ export function request(name: string, username: string, args: ArgsMapping, next:
...
+      args['user_id'] = user.user_id;
```
Comment (emmiegit) [inline backend/src/services/pagereq/index.ts:533]: If you have a static key you can do `args.user_id`. Also why is this snake_case?

File: backend/src/services/pagereq/index.ts [hunk 11]
```diff
@@ -456,32 +521,35 @@ export function request(name: string, username: string, args: ArgsMapping, next:
...
-    if (name === 'changePage') changePage(username, args, next);
-    else if (name === 'removeEditLock') removeEditLock(username, args, next);
-    else if (name === 'beginEditPage') beginEditPage(username, args, next);
-    else if (name === 'voteOnPage') voteOnPage(username, args, next);
-    else if (name === 'tagPage') tagPage(username, args, next);
+    if (name === 'changePage') changePage(user, args, next);
+    else if (name === 'removeEditLock') removeEditLock(user, args, next);
+    else if (name === 'beginEditPage') beginEditPage(user, args, next);
+    else if (name === 'voteOnPage') voteOnPage(user, args, next);
+    else if (name === 'tagPage') tagPage(user, args, next);
```
Comment (emmiegit) [inline backend/src/services/pagereq/index.ts]: Use a `switch` here

File: backend/src/services/user/index.ts [hunk 12]
```diff
@@ -33,6 +34,7 @@ const randomBytesPromise = promisify(randomBytes);
 // represents a user
 // TODO: figure out the best way to incorporate stats into this
 export class User {
+  public static systemUser: Nullable<User> = null;
```
Comment (emmiegit) [inline backend/src/services/user/index.ts]: This should either be initialized on class creation or stored per-wiki. Having a potentially null constant "system user" is very strange

File: backend/src/services/user/index.ts [hunk 13]
```diff
@@ -44,8 +46,15 @@ export class User {
     public city: Nullable<string>,
     public avatar: string,
     public gender: Nullable<string>,
+    public role: Role
   ) {}
 
+  // tell if the user has permission to do something
+  hasPermission(permname: string): boolean {
```
Comment (emmiegit) [inline backend/src/services/user/index.ts]: You can make the permission key a `type` to avoid possible typos in the string

File: backend/src/services/user/index.ts [hunk 14]
```diff
@@ -103,30 +113,26 @@ export class User {
...
-
-    return user_object.validate(password);
-  }
-
   // add a new user to the database
   // NOTE: number returned is an error code
   static async createNewUser(username: string,
                              email: string,
                              password: string,
+                             role: Role | number,
```
Comment (emmiegit) [inline backend/src/services/user/index.ts:122]: If you only want an id call it `roleId` and take `number`. A consumer with a `Role` can always just do `role.id`

File: backend/src/services/user/permissions.ts [hunk 15]
```diff
@@ -0,0 +1,129 @@
...
+  { permission: { name: "lockPages", display_name: "Lock Pages" }, value: false },
+  { permission: { name: "modifyLockedPages", display_name: "Modify Locked Pages (Every action they can do normally, can be done to locked pages)" }, value: false },
+
+  { permission: { name: "createEditRoles", display_name: "Create and Edit Roles" }, value: false },
+  { permission: { name: "promoteRoles", display_name: "Promote Users to Roles" }, value: false },
+];
+
+
+
+export const NumPermissions = DefaultPermissions.length;
```
Comment (emmiegit) [inline backend/src/services/user/permissions.ts]: Why is this necessary? Why can't I just do `DefaultPermissions.length`

Comment (emmiegit) [reply backend/src/services/user/permissions.ts]: Also constants should be `SCREAM_SNAKE_CASE` -- this case makes them look like class names

File: backend/src/services/user/permissions.ts [hunk 16]
```diff
@@ -0,0 +1,129 @@
...
+    this.permissions = DefaultPermissions.map(p => Object.assign({}, p)); // deep copy
+  }
+
+  // get a permset from a stored number
+  static fromNumber(value: number): Permset {
+    //console.log(`=== BITSET is ${(value >>> 0).toString(2)}`);
+    let permset = new Permset();
+    for (let i = 0; i < NumPermissions; i++) {
+      permset.permissions[i].value = getLargenumVal(value, i); 
+      //console.log(`PERMISSIONS - Value for ${permset.permissions[i].permission.name} is ${getLargenumVal(value, i)}`);
```
Comment (emmiegit) [inline backend/src/services/user/permissions.ts]: Remove commented code
```suggestion
```

File: backend/src/services/user/permissions.ts [hunk 17]
```diff
@@ -0,0 +1,129 @@
...
+    let permission;
+
+
+    for (let i = 0; i < NumPermissions; i++) {
+      permission = this.permissions[i];
+
+      if (this.permissions[i].value) {
+
+        value |= (1 << i); 
+
```
Comment (emmiegit) [inline backend/src/services/user/permissions.ts]: Why is there so much random whitespace? Same with lines 66-67 and 48-50

File: backend/src/services/user/permissions.ts [hunk 18]
```diff
@@ -0,0 +1,129 @@
...
+      }
+    }
+
+    return value;
+  }
+
+  // get the value of a permissions
+  hasPermission(permname: string): boolean {
+    for (let i = 0; i < NumPermissions; i++) {
+      let permission = this.permissions[i];
```
Comment (emmiegit) [inline backend/src/services/user/permissions.ts]: If it's not reassigned you should use `const`

Comment (emmiegit) [reply backend/src/services/user/permissions.ts]: Also use `const permission of DefaultPermissions` instead of index based iteration

File: backend/src/services/user/role.ts [hunk 19]
```diff
@@ -80,8 +65,40 @@ export class Role {
...
+    let rolecheck = await query("SELECT * FROM Roles WHERE role_name=$1;", [rolename]);
+    if (rolecheck.rowCount > 0) {
+      if (return_role) return Role.fromRow(rolecheck.rows[0]);
+      else return ErrorCode.ROLE_EXISTS;
+    }
+
+    if (permset) {
+      if (permset instanceof Permset) {
+        permset = permset.getNumber();
+      }
```
Comment (emmiegit) [inline backend/src/services/user/role.ts:96]: This permset checking pattern keeps reoccurring, there should just be a helper utility so you can get the right type right at the start

File: backend/gulpfile.js (hunk 2 above)
Comment (notgull) [reply backend/gulpfile.js:27]: It compiles for me

File: backend/dist/app (hunk 3 above)
Comment (notgull) [reply backend/dist/app:1]: I had to remove `dist` at one point and recreate the symlink. Looks like I used a different way of getting the current folder.

File: backend/src/services/metadata/autocreate-404.ts (hunk 6 above)
Comment (notgull) [reply backend/src/services/metadata/autocreate-404.ts:67]: I should rename `autocreate-404.ts`, because it's becoming less of "just copy over some initial pages" and more of "set up essential users, roles and variables that the system needs to function". I want to keep the `systemUser` stored somewhere, so rest assured that this is not "ad-hoc".

File: backend/src/services/pagereq/index.ts (hunk 7 above)
Comment (notgull) [reply backend/src/services/pagereq/index.ts]: I generally like being explicit. I like to specify that this is the opposite of `user.hasPermission` rather than "take the opposite of `user` then call `hasPermission` on it"

Comment (emmiegit) [reply backend/src/services/pagereq/index.ts]: This isn't any more explicit though, order of operations prevents this from being opposite of user, to get that you'd need `(!user).hasPermission`

File: backend/src/services/user/index.ts (hunk 12 above)
Comment (notgull) [reply backend/src/services/user/index.ts]: This is intended to be the `systemUser` created or loaded in `autocreate-404`. Although I think it should be replaced with something to the tune of `public static systemUserName: string = "system";` because IMO it's a better way of storing the system user.

File: backend/src/services/metadata/autocreate-404.ts (hunk 6 above)
Comment (emmiegit) [reply backend/src/services/metadata/autocreate-404.ts:67]: I would suggest making it a script since it does a lot of things that don't make much sense as part of the metadata service

File: backend/src/services/user/permissions.ts (hunk 15 above)
Comment (notgull) [reply backend/src/services/user/permissions.ts]: 1). `DefaultPermissions.length` is a somewhat time-consuming operation for something that will be theoretically called three or four times per `pagereq` operation.

2). Done

File: backend/src/services/user/permissions.ts (hunk 16 above)
Comment (notgull) [reply backend/src/services/user/permissions.ts]: Noting that I did this in the last commit.

File: backend/src/services/metadata/autocreate-404.ts (hunk 6 above)
Comment (notgull) [reply backend/src/services/metadata/autocreate-404.ts:67]: You mean, moving it to the root `src` dir?

File: backend/src/services/user/permissions.ts (hunk 15 above)
Comment (emmiegit) [reply backend/src/services/user/permissions.ts]: What? I double checked and yeah, it's just an integer storing the current length. Like this needs to happen anyways on the lower level because it needs to know how much of its buffer it has used.

This is a `O(1)` operation, not `O(n)`. Accessing it isn't any more expensive than any other random field.

File: backend/src/services/pagereq/index.ts (hunk 10 above)
Comment (notgull) [reply backend/src/services/pagereq/index.ts:533]: Main reason is because `user_id` is snakecase all over the code and I haven't gotten around to fixing that yet

File: backend/src/services/metadata/autocreate-404.ts (hunk 6 above)
Comment (emmiegit) [reply backend/src/services/metadata/autocreate-404.ts:67]: Like there's that bin directory that does a bunch of misc setup stuff.

File: backend/src/services/user/index.ts (hunk 12 above)
Comment (emmiegit) [reply backend/src/services/user/index.ts]: Yeah I'm not a fan of lazy auto loading controlled by a side module like that. It's not clear at all how this is meant to be initialized or anything.

File: backend/src/services/user/permissions.ts [hunk 20]
```diff
@@ -99,10 +110,8 @@ export class Permset {
 
   // get the value of a permissions
   hasPermission(permname: string): boolean {
```
Comment (emmiegit) [inline backend/src/services/user/permissions.ts]: You want `permname: PermissionName` not `string`

File: backend/src/services/metadata/autocreate-404.ts (hunk 6 above)
Comment (notgull) [reply backend/src/services/metadata/autocreate-404.ts:67]: I like the design idea of the code being able to set itself up.

File: backend/src/services/user/permissions.ts [hunk 21]
```diff
@@ -18,6 +18,19 @@
...
+                             "createPages" |
+                             "voteOnPages" |
+                             "tagPages" |
+                             "revertRevisions" |
+                             "deletePages" |
+                             "ratePages" |
+                             "lockPages" |
+                             "modifyLockedPages" |
+                             "createEditRoles" |
+                             "promoteRoles";
```
Comment (emmiegit) [inline backend/src/services/user/permissions.ts:31]: btw I've been doing
```ts
export type PermissionName =
  | 'editPages'
  | 'createPages'
  | 'voteOnPages'
  | ...
```
since it looks cleaner

File: backend/src/services/user/permissions.ts [hunk 22]
```diff
@@ -18,6 +18,19 @@
...
+*/ // Noting that I tried this but it threw an error with Object.assign()
```
Comment (emmiegit) [inline backend/src/services/user/permissions.ts]: What was the issue?

File: backend/src/services/metadata/autocreate-404.ts (hunk 6 above)
Comment (emmiegit) [reply backend/src/services/metadata/autocreate-404.ts:67]: That's fine, then there should be an explicit initialization path instead of it being scattered throughout.

Also consider there are multiple kinds of initialization here. There's setting up internal data structures when it begins running, and also setting up and migrating the database.

Comment (notgull) [reply backend/src/services/metadata/autocreate-404.ts:67]: I think it also might be good to merge `setup_db` into this, then.

I'll probably take care of that once I've wrangled out the web server.

File: backend/src/services/user/role.ts (hunk 19 above)
Comment (notgull) [reply backend/src/services/user/role.ts:96]: It involves something different each time.

File: backend/src/index.ts (hunk 4 above)
Comment (notgull) [reply backend/src/index.ts:188]: Will probably find a better way of doing this once I refactor the web API

File: backend/src/services/user/index.ts (hunk 14 above)
Comment (notgull) [reply backend/src/services/user/index.ts:122]: I like the option of being able to pass in either a `Role` or a role id depending on what you have on hand

File: backend/src/services/user/index.ts [hunk 23]
```diff
@@ -86,11 +96,12 @@ export class User {
                     row.about,
                     row.city,
                     row.avatar,
-                    row.gender);
+                    row.gender,
+                    await Role.loadById(row.role_id));
```
Comment (emmiegit) [inline backend/src/services/user/index.ts:100]: This `fromRow` pattern is basically a deserializer, and I don't like the idea of mixing a query into there. Imo the database layer should handle models cleaner so this method isn't needed.

Anyways the larger issues here are:
* users can only have one role?
* just provide the list of role ids in an attribute - if somebody wants the role objects they can fetch them themselves

Comment (notgull) [reply backend/src/services/user/index.ts:100]: 1). You're right - tenants are going to have to be worked into this. However, at the moment, I think I'll leave it like this, given our one-tenant setup.

2). The main reason it's like this setup is similar to why there is a similar on-load setup in `metadata`- one of the main uses for user objects is for their accompanying `Role` object, which is called in `hasPermission`. I think it would make the code messier to have it loaded manually.

Comment (emmiegit) [reply backend/src/services/user/index.ts:100]: 1. We should start adding TODOs in these sections.

2. I'm fine loading the `Role` object, my main criticism is that this should exist where the user querying is happening, not in the deser.

**Review by emmiegit — APPROVED** [review]
+1 after unresolved comments

File: backend/src/services/user/index.ts (hunk 23 above)
Comment (notgull) [reply backend/src/services/user/index.ts:100]: 1). ok

2). `fromRow` is mostly just to simplify functionality for converting an SQL row into an object
